Use tuple inference for closures

roxelo opened the pull request:
The main motivation behind this change is to allow the compiler to decide the number of captures after capture analysis instead of before capture analysis. T

This is required to support RFC-2229, since the number of captures will not be know before capture analysis.

Closes rust-lang/project-rfc-2229#4

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sexxi-goose/rust/23)
<!-- Reviewable:end -->

roxelo committed:
Replace tuple of infer vars for upvar_tys with single infer var

This commit allows us to decide the number of captures required after
completing capture ananysis, which is required as part of implementing
RFC-2229.

Co-authored-by: Aman Arora <me@aman-arora.com>
Co-authored-by: Jenny Wills <wills.jenniferg@gmail.com>

arora-aman commented on compiler/rustc_middle/src/ty/outlives.rs:
```diff
@@ -89,8 +89,13 @@ fn compute_components(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, out: &mut SmallVec<[Compo
-                for upvar_ty in substs.as_closure().upvar_tys() {
-                    compute_components(tcx, upvar_ty, out);
+                let tupled_ty = substs.as_closure().tupled_upvars_ty();
+                if matches!(tupled_ty.kind(), ty::Tuple(_)) {
```
This is my bad, this can just be `substs.as_closure().is_valid()`

arora-aman commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -1730,11 +1736,31 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
-            ty::Closure(_, ref substs) => substs.as_closure().upvar_tys().collect(),
+            ty::Closure(_, ref substs) => {
+                let ty = self.infcx.shallow_resolve(substs.as_closure().tupled_upvars_ty());
+                if let ty::Infer(ty::TyVar(_)) = ty.kind() {
```
I think you can just do `constituent_types_for_ty(ty)`, but don't think that's the issue rn

arora-aman commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -1730,11 +1736,31 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
-                substs.as_generator().upvar_tys().chain(iter::once(witness)).collect()
+                let ty = self.infcx.shallow_resolve(substs.as_generator().tupled_upvars_ty());
+                if let ty::Infer(ty::TyVar(_)) = ty.kind() {
+                    // Not yet resolved.
```
same as above

arora-aman commented on compiler/rustc_trait_selection/src/opaque_types.rs:
```diff
@@ -716,24 +720,36 @@ where
+                let ty = self.infcx.shallow_resolve(substs.as_closure().tupled_upvars_ty());
+                if let ty::Infer(ty::TyVar(_)) = ty.kind() {
+                    // Not yet resolved.
+                    ty.super_visit_with(self);
```
Not imp rn but I'd just call visit_ty, here.

arora-aman commented on compiler/rustc_trait_selection/src/opaque_types.rs:
```diff
@@ -716,24 +720,36 @@ where
+                let ty = self.infcx.shallow_resolve(substs.as_generator().tupled_upvars_ty());
+                if let ty::Infer(ty::TyVar(_)) = ty.kind() {
+                    // Not yet resolved.
+                    ty.super_visit_with(self);
```
same

arora-aman commented on compiler/rustc_trait_selection/src/traits/wf.rs:
```diff
@@ -526,10 +526,7 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
-                        // FIXME(eddyb) add the type to `walker` instead of recursing.
-                        self.compute(upvar_ty.into());
-                    }
+                    self.compute(substs.as_closure().tupled_upvars_ty().into());
```
It feels like at this point we would've done capture analysis, based on the comments, but not sure.

Also how `ty::Tuple` is dealt with within this code seems to be different than what was done here before this change.

arora-aman commented on compiler/rustc_trait_selection/src/traits/query/dropck_outlives.rs:
```diff
@@ -110,7 +110,12 @@ pub fn trivial_dropck_outlives<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> bool {
         ty::Tuple(ref tys) => tys.iter().all(|t| trivial_dropck_outlives(tcx, t.expect_ty())),
         ty::Closure(_, ref substs) => {
-            substs.as_closure().upvar_tys().all(|t| trivial_dropck_outlives(tcx, t))
+            if !substs.as_closure().is_valid() {
```
This can just be `trivial_dropck_outlives(tcx, substs.as_closure().tupled_upvar_tys())`

arora-aman commented on compiler/rustc_trait_selection/src/traits/wf.rs:
```diff
@@ -526,10 +526,7 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
                     // only inspects the upvar types).
                     walker.skip_current_subtree(); // subtree handled below
-                    for upvar_ty in substs.as_closure().upvar_tys() {
-                        // FIXME(eddyb) add the type to `walker` instead of recursing.
```
leave this comment here

arora-aman commented on compiler/rustc_typeck/src/check/closure.rs:
```diff
@@ -81,19 +81,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                })
-            }));
+        let tupled_upvars_ty = self.infcx.next_ty_var(TypeVariableOrigin {
+            // FIXME(eddyb) distinguish upvar inference variables from the rest.
```
I feel like this isn't valid anymore, individual inference variable for each upvar, isn't a thing anymore.

arora-aman committed:
Always return tupled_upvar_tys for Closure/Generator consituent tys

Depending on if upvar_tys inferred or not, we were returning either an
inference variable which later resolves to a tuple or else the upvar tys
themselves

Co-authored-by: Roxane Fruytier <roxane.fruytier@hotmail.com>

roxelo commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -1307,6 +1307,26 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         let mut next_code = Some(&obligation.cause.code);
+        // By introducing a tuple into the upvar types, the first item in the obligation chain
+        // can be of Tuple type instead of the generator that captured the type. We want to catch for
+        // this case and skip it.
```
This does the trick for now, but it's probably not what we want to move forward with. The tuple of upvar type should not be the first item in the obligation chain, probably more in the middle. We probably want to change how the obligation chain is created. I haven't figured out where in the code that is done yet though

arora-aman committed:
traits diagnostics: Don't print closure/generator upvar_tys tuple

Co-authored-by: Roxane Fruytier <roxane.fruytier@hotmail.com>

roxelo committed:
Address comments
